credentials/tls/certprovider/meshca/plugin.go

```diff
@@ -0,0 +1,280 @@
+	dist     *certprovider.Distributor // Holds the key material.
+	logger   *grpclog.PrefixLogger     // Plugin instance specific prefix.
+	backoff  func(int) time.Duration   // Exponential backoff.
+	doneFunc func()                    // Notify the builder when done.
+}
+
+// providerParams wraps params passed to the provider plugin at creation time.
+type providerParams struct {
+	// This ClientConn to the MeshCA server is owned by the builder.
+	cc       *grpc.ClientConn
+	cfg      *pluginConfig
+	opts     certprovider.Options
+	backoff  func(int) time.Duration
+	doneFunc func()
+}
+
+func newProviderPlugin(params providerParams) *providerPlugin {
+	ctx, cancel := context.WithCancel(context.Background())
+	p := &providerPlugin{
+		cancel:   cancel,
+		cc:       params.cc,
+		cfg:      params.cfg,
+		opts:     params.opts,
+		backoff:  params.backoff,
+		doneFunc: params.doneFunc,
+		dist:     certprovider.NewDistributor(),
+	}
+	p.logger = prefixLogger((p))
+	p.logger.Infof("plugin created")
+	go p.run(ctx)
+	return p
+}
+
+func (p *providerPlugin) KeyMaterial(ctx context.Context) (*certprovider.KeyMaterial, error) {
+	return p.dist.KeyMaterial(ctx)
+}
+
+func (p *providerPlugin) Close() {
+	p.logger.Infof("plugin closed")
+	p.dist.Stop()
+	p.cancel()
+	p.doneFunc()
+}
+
+// run is a long running goroutine which periodically sends out CSRs to the
+// MeshCA, and updates the underlying Distributor with the new key material.
+func (p *providerPlugin) run(ctx context.Context) {
+	// We need to start fetching key material right away. The next attempt will
+	// be triggered by the timer firing.
+	certValidity := p.updateKeyMaterial(ctx)
+	if certValidity == 0 {
+		return
+	}
+
+	for {
+		// We request a certificate whose validity is usually twice as much as
+		// the grace period. But the server is free to return a certificate with
+		// whatever validity time it deems right.
+		refreshAfter := p.cfg.certGraceTime
+		if refreshAfter > certValidity {
+			refreshAfter = certValidity
+		}
+		timer := time.NewTimer(refreshAfter)
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+			return
+		case <-timer.C:
+			certValidity = p.updateKeyMaterial(ctx)
+		}
+	}
+}
+
+// updateKeyMaterial generates a CSR and attempts to get it signed from the
+// MeshCA. It retries with an exponential backoff till it succeeds or the
+// deadline specified in ctx expires. Once it gets the CSR signed from the
+// MeshCA, it updates the Distributor with the new key material.
+//
+// It returns the amount of time the new certificate is valid for, or zero if
+// the context expires.
+func (p *providerPlugin) updateKeyMaterial(ctx context.Context) time.Duration {
+	client := meshpb.NewMeshCertificateServiceClient(p.cc)
+	retries := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return 0
+		default:
+		}
+
+		if retries != 0 {
+			timer := time.NewTimer(p.backoff(retries))
+			select {
+			case <-timer.C:
+			case <-ctx.Done():
+				if !timer.Stop() {
+					<-timer.C
+				}
```

Comment:
This stuff is only needed if you're going to be reusing `timer`.  Just `timer.Stop(); return` is fine.